README.md
[![Build Status](https://travis-ci.org/n3okill/enfspatch-promise.svg)](https://travis-ci.org/n3okill/enfspatch-promise)
[![AppVeyour status](https://ci.appveyor.com/api/projects/status/hqcy30ki773s6rbw?svg=true)](https://ci.appveyor.com/project/n3okill/enfspatch-promise)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/f2c315ec874a4703998eadf1bf6b585c)](https://www.codacy.com/app/n3okill/enfspatch-promise)
[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=64PYTCDH5UNZ6)

[![NPM](https://nodei.co/npm/enfspatch-promise.png)](https://nodei.co/npm/enfspatch-promise/)

enfspatch-promise
=================
Additional methods and patches for node fs module

**enfs** stands for [E]asy [N]ode [fs]

This module is intended to work as a sub-module of [enfs](https://www.npmjs.com/package/enfs-promise)


Description
-----------
This module is just a wrapper for [enfspath](https://www.npmjs.com/package/enfspath) to enable the use of promises.
  
Usage
-----
`enfspatch-promise` is a drop-in replacement for native `fs` module, you just need to include
it instead of the native module.

Use this
```js
    const enfs = require("enfspatch-promise");
```

All async methods finish with 'P', example:
open => openP

```js
    enfs.openP("/path","flags").then(function(fd){
        console.log(fd);
    });
```

instead of

```js
    const fs = require("fs"); //You don't need to do this anymore
```

and all the methods from native fs module are available fir use with promises

Errors
------
All the methods follows the node culture.
- Async: Every async method returns an Error in the first callback parameter
- Sync: Every sync method throws an Error.


License
-------

Creative Commons Attribution 4.0 International License

Copyright (c) 2017 Joao Parreira <joaofrparreira@gmail.com> [GitHub](https://github.com/n3okill)

This work is licensed under the Creative Commons Attribution 4.0 International License. 
To view a copy of this license, visit [CC-BY-4.0](http://creativecommons.org/licenses/by/4.0/).




CHANGELOG.md
1.0.2 / 2017-04-11
  * arguments length correction
  
1.0.0 / 2017-02-21
==================
  * first release


test/open.js
/**
 * @project enfspatch-promise
 * @filename open.js
 * @description tests for enfspatch-promise
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.1
 */

/* global describe, require, it, __filename */

"use strict";

const enfs = require("../");

describe("enfspatch > Open", function () {
    it("should open an existing file async", function () {
        return enfs.openP(__filename, "r").then(function (fd) {
            (typeof fd === "undefined").should.be.equal(false);
            return enfs.closeP(fd);
        }).should.be.fulfilled();
    });
    it("should fail to open non-existing file async", function () {
        return enfs.openP("invalid file path", "r").catch(function (err) {
            err.should.be.instanceOf(Error);
            err.code.should.be.equal("ENOENT");
        });
    });
});


test/stats.js
/**
 * @project enfspatch
 * @filename stats.js
 * @description tests for enfspatch
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.1
 */

/* global describe, require, it, __filename */

"use strict";

const enfs = require("../");

describe("enfspatch > Stats", function () {
    it("enfs should use the same stats constructor as fs module", function () {
        return enfs.statP(__filename).then(function (statFs) {
            statFs.isFile().should.be.equal(true);
        });
    });
});


test/rename.js
/**
 * @project enfspatch-promise
 * @filename rename.js
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2017 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 13-02-2017
 * @version 0.1.1
 * @description tests for enfspatch-promise [rename]
 */

/* global describe, it, before, after*/

"use strict";

const nodePath = require("path");
const nodeOs = require("os");
const rimraf = require("rimraf");
const cwd = process.cwd();
const enFs = require("../");

describe("enfsPatch-promise > rename", function () {
    const tmpPath = nodePath.join(nodeOs.tmpdir(), "enfspatchpromiserenamefile");
    before(function () {
        try {
            enFs.mkdirSync(tmpPath);
            process.chdir(tmpPath);
        } catch (err) {
            if (err.code === "EXIST") {
                rimraf.sync(tmpPath);
                enFs.mkdirSync(tmpPath);
            }
        }
    });
    after(function () {
        process.chdir(cwd);
        rimraf.sync(tmpPath);
    });
    describe("> async", function () {
        describe("try one file", function () {
            let fileALock;
            let fileA, fileB, fileARenamed, fileBRenamed, fsRename, fsRenameP;
            before(function () {
                fsRename = enFs.rename;
                enFs.rename = ((function (rename) {
                    return function (from, to, callback) {
                        if (fileALock === from) {
                            let err = new Error("File Locked by system.");
                            err.code = "EPERM";
                            return callback(err);
                        }
                        rename(from, to, callback);
                    };
                })(enFs.rename));
                fsRenameP = enFs.renameP;
                enFs.renameP = ((function () {
                    return function (from, to) {
                        return new Promise(function (resolve, reject) {
                            enFs.rename(from, to, (err) => {
                                if (err) {
                                    reject(err);
                                } else {
                                    resolve();
                                }
                            });
                        });
                    };
                })(enFs.renameP));
                fileA = nodePath.join(tmpPath, "a.txt");
                fileARenamed = nodePath.join(tmpPath, "aRename.txt");
                fileB = nodePath.join(tmpPath, "b.txt");
                fileBRenamed = nodePath.join(tmpPath, "bRenamed.txt");
            });
            after(function () {
                enFs.rename = fsRename;
                enFs.renameP = fsRenameP;
            });
            it("should write files and lock fileA", function () {
                return enFs.writeFileP(fileA, "data", "utf8")
                    .then(function () {
                        fileALock = fileA;
                        return enFs.writeFileP(fileB, "data", "utf8");
                    }).should.be.fulfilled();
            });
            it("should rename fileB and fail on fileA", function () {
                return enFs.renameP(fileB, fileBRenamed)
                    .then(function () {
                        return enFs.statP(fileBRenamed);
                    })
                    .then(function (stats) {
                        stats.isFile().should.be.equal(true);
                        return enFs.renameP(fileA, fileARenamed);
                    }).should.be.rejectedWith(Error, {code: "EPERM"});
            });
            it("should rename fileA only after unlocked", function () {
                return enFs.renameP(fileA, fileARenamed)
                    .catch(function (err) {
                        err.code.should.be.equal("EPERM");
                        fileALock = null;
                        return enFs.renameP(fileA, fileARenamed);
                    })
                    .then(function () {
                        return enFs.statP(fileARenamed);
                    })
                    .then(function (stats) {
                        stats.isFile().should.be.equal(true);
                    });
            });
        });
        describe("a lot of files", function () {
            const locks = [];
            let fsRename = enFs.rename;
            const fsRenameP = enFs.renameP;
            let num, paths;
            before(function () {
                num = 500;
                paths = new Array(num);
                fsRename = enFs.rename;
                enFs.rename = ((function (rename) {
                    return function (from, to, callback) {
                        if (locks.indexOf(from) !== -1) {
                            let err = new Error("File Locked by system.");
                            err.code = "EPERM";
                            return callback(err);
                        }
                        rename(from, to, callback);
                    };
                })(enFs.rename));
                enFs.renameP = ((function () {
                    return function (from, to) {
                        return new Promise(function (resolve, reject) {
                            enFs.rename(from, to, (err) => {
                                if (err) {
                                    reject(err);
                                } else {
                                    resolve();
                                }
                            });
                        });
                    };
                })(enFs.renameP));
            });
            after(function () {
                enFs.rename = fsRename;
                enFs.renameP = fsRenameP;
            });

            function randomIntInc(low, high) {
                return Math.floor(Math.random() * (high - low + 1) + low);
            }

            function getRandomArray(low, high, size) {
                let numbers = [];
                for (let i = 0; i < size; i++) {
                    numbers.push(randomIntInc(low, high));
                }
                return numbers.filter((v, i, a) => a.indexOf(v) === i);
            }


            it("should write files and lock some", function () {
                let ps = [];
                this.timeout(10000);
                for (let i = 0; i <= num; i++) {
                    paths[i] = nodePath.join(tmpPath,"file-" + i.toString());
                    ps.push(enFs.writeFileP(paths[i], "data", "utf8"));
                }
                return Promise.all(ps).then(function () {
                    let numbers = getRandomArray(0, num, randomIntInc(1, num));
                    while (numbers.length) {
                        locks.push(paths[numbers.shift()]);
                    }
                });
            });
            it("should rename not locked files", function () {
                let renamed = 0, notRenamed = 0;
                let locked = locks.length;
                let ps = paths.map((path) =>
                    enFs.renameP(path, path + ".renamed")
                        .then(function () {
                            renamed++;
                        })
                        .catch(function (err) {
                            err.code.should.be.equal("EPERM");
                            notRenamed++;
                        })
                );
                return Promise.all(ps).then(function () {
                    notRenamed.should.be.equal(locked);
                    renamed.should.be.equal(paths.length - locked);
                });
            });
            it("should rename locked files after unlock", function () {
                let locked = locks.length;
                let renamed = 0;
                let tmpLocks = [];
                for (let i = 0; i < locked; i++) {
                    tmpLocks[i] = locks[i];
                }
                let ps = locks.map((path) =>
                    enFs.renameP(path, path + ".renamed")
                        .catch(function (err) {
                            err.code.should.be.equal("EPERM");
                        })
                );
                return Promise.all(ps).then(function () {
                    renamed.should.be.equal(0);
                    while (locks.length) {
                        locks.shift();
                    }
                    let ps1 = tmpLocks.map((path) =>
                        enFs.renameP(path, path + ".renamed")
                            .then(function () {
                                renamed++;
                            })
                    );
                    Promise.all(ps1).then(function () {
                        renamed.should.be.equal(locked);
                    })
                })
            });
        });
    });
});


lib/enfspatch-promise.js
/**
 * @project enfspatch-promise
 * @filename enfspatch-promise.js
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2017 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 21-02-2017
 * @version 0.0.1
 * @description
 */

"use strict";

const NodeAssert = require("assert");
const enfsAddins = require("enfsaddins-promise");

function promisify(fs) {
    const fns = Object.keys(fs);

    const result = fs;
    fns.forEach((key) => {
        if (key in fs && (key.length > 3 && (key.slice(-4) !== "Sync" && key.slice(-5) !== "Stream" && key !== "Stats")) && typeof fs[key] === "function") {
            result[`${key}P`] = createPromise(fs[key]);
        }
    });
    return result;
}

module.exports = enfsAddins(promisify(require("enfspatch")));
module.exports.promisify = promisify;


function createPromise(fn) {
    NodeAssert.strictEqual(typeof fn, "function", "fn should be a function.");
    NodeAssert.notStrictEqual(typeof fn, "undefined", "fn can't be undefined");
    NodeAssert.notStrictEqual(null, fn, "fn can't be null");
    return function () {
        const self = this;
        const args = Array.from(arguments);
        const len = fn.length;
        if (!!len && args.length < len) {
            while (args.length < (len-1)) {
                args.push(void 0);
            }
        }
        return new Promise(function (resolve, reject) {
            args.push(function (err, result) {
                if (err) {
                    reject(err);
                } else {
                    resolve(result);
                }
            });
            fn.apply(self, args);
        });
    };
}

